### ISSUE TITLE
Bug: DevTools calls arbitrary generators which may be stateful

### ISSUE THREAD

**Issue by gaearon** [facebook/react#19726]
```js
function foo*() {
  yield 1;
  yield 2;
}

let gen = foo()
```

Currently if you put `gen` into state or props and then open this component in DevTools, it will consume that generator while trying to format it. So `gen.next()` will give you `{ done: true }` next time you call it.

This happens here:

https://github.com/facebook/react/blob/60ba723bf78b9a28f60dce854e88e206fab52301/packages/react-devtools-shared/src/utils.js#L616-L623

I think that maybe we should treat iterables differently if they *return themselves* as an iterator. Since that means they're likely stateful and it's not ok to iterate over them.

We detect iterables here (DevTools terminology is wrong btw, it should be `iterable` rather than `iterator`):

https://github.com/facebook/react/blob/60ba723bf78b9a28f60dce854e88e206fab52301/packages/react-devtools-shared/src/utils.js#L438-L439

I think maybe we could split this into `iterable` and `opaque_iterable`, and make sure none of the codepaths attempt to traverse `opaque_iterable` or pass it to something that would consume it (e.g. `Array.from`).

We could detect it based on `data[Symbol.iterator]() === data` — that clearly signals the iterable is its own iterator (which is the case for generators), and therefore it's not OK for DevTools to consume it.

Maybe some other heuristic could work. But overall, the goal is that `Map` and friends is still being iterated over, but an arbitrary generator is not.

**Comment by eps1lon** [comment]
This is similar to https://github.com/facebook/react/issues/19660 where the same line of code causes devtools to crash with infinite generators. 

I wasn't able to reproduce that devtools calling stateful generators was an issue though I definitely expected it to be a problem.

**Comment by todortotev** [comment]
@gaearon I could give a shot on this!

**Comment by abelmark** [comment]
@gaearon I would love to give this a go.

**Comment by goanpixie** [comment]
@gaearon : I would like to address this bug

**Comment by eps1lon** [comment]
@todortotev Go for it :+1: Please let us know if you no longer intend to work on it.

**Comment by todortotev** [comment]
> @todortotev Go for it 👍 Please let us know if you no longer intend to work on it.

On it!

**Comment by kambleaa007** [comment]
just let us know branch you are putting your efforts, more guys would like to contribute, we all can Go for it 👍

**Comment by omarsy** [comment]
@eps1lon I agree with you both issues are related.  if we choose to not extract generator. We’re not gonna have the issue anymore

**Comment by todortotev** [comment]
Just a note, I'm nearly done! Sorry for being slow it just was my first attempt!

**Comment by engprodigy** [comment]
@todortotev  [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators#Iterables] This can help

**Comment by bvaughn** [comment]
This should be resolved by #19831 once it lands.
